docs/dev/implementation-notes.md

```diff
@@ -0,0 +1,160 @@
+
+- *Vendorability.* array-api-compat should be [vendorable](vendoring). This
+  means that, for instance, all imports in the library are relative imports.
+  No code in the package specifically references the name `array_api_compat`
+  (we also support renaming the package to something else).
+  Vendorability support is tested in `tests/test_vendoring.py`.
+
+- *Pure Python.* To make array-api-compat as easy as possible to add as a
+  dependency, the code is all pure Python.
+
+- *Minimal Wrapping Only.* The wrapping functionality is minimal. This means
+  that if something is difficult to wrap using pure Python, or if trying to
+  support some array API behavior would require a significant amount of code,
+  we prefer to leave the behavior as an upstream issue for the array library,
+  and [document it as a known difference](../supported-array-libraries.md).
+
+  This also means that we do not at this point in time implement anything
+  other than wrappers for functions in the standard, and basic [helper
+  functions](../helper-functions.md) that would be useful for most users of
+  array-api-compat. The addition of functions that are not part of the array
+  API standard is currently out-of-scope for this package.
+
+- *No Monkey Patching.* `array-api-compat` should not attempt to modify anything
+  about the underlying library. It is a *wrapper* library only.
+
+- *No modifying the array object.* Due to the above restrictions, the array
+  (or tensor) object of the array library cannot be modified. Any behavior
+  that is built-in to the array object, such as the behavior of [array
+  methods](https://data-apis.org/array-api/latest/API_specification/array_object.html)
+  is therefore left unwrapped. Users can workaround issues by using
+  corresponding [elementwise
+  functions](https://data-apis.org/array-api/latest/API_specification/elementwise_functions.html)
+  instead of operators, and using the [helper
+  functions](../helper-functions.md) provided by array-api-compat instead of
+  methods like `to_device`.
+
+- *Avoid Restricting Non-Standard Behavior.* All array libraries have
+  functions and behaviors that are outside of the scope of what is specified
+  by the standard. These behaviors should be left intact whenever possible,
+  unless the standard explicitly disallows something. This means
+
+  - All namespaces are *extended* with wrapper functions. You may notice the
+    extensive use of `import *` in various files in `array_api_compat`. While
+    this would normally be questionable, this is the [one actual legitimate
+    use-case for `import *`](https://peps.python.org/pep-0008/#imports), to
+    re-export names from an external namespace.
+
+  - All wrapper functions pass `**kwargs` through to the wrapped function.
+
+  The onus is on users of array-api-compat to ensure they are only using
+  portable array API behavior, e.g., by testing against [array-api-strict](array-api-strict).
+
+
+## Avoiding Code Duplication
+
```

Comment:
This section may want to add a mention of JAX and Dask, even to just mention that they exist. Or use "for example" when namespaces are explicitly listed